### PR TITLE
Implement ERC20Cogateway::withdraw()

### LINKED ISSUE #273: Implement ERC20Cogateway::withdraw
Implement `withdraw` method in ERC20Cogateway
- Burn the Utility token from the withdrawer address.
```
    ERC20Cogateway.withdraw
        - utility token address
        - amount
        - beneficiary address
        - gas price
        - gas limit
```
    
Emit event: WithdrawIntentDeclared
```
WithdrawIntentDeclared
    - utility token address
    - amount
    - nonce
    - beneficiary
    - fee gas price
    - fee gas limit
    - withdrawer
```

### PR DESCRIPTION
**Pull request by Vindeep07**
Fixes #273 

- Implemented ERC20Cogateway::withdraw()
- Added unit tests

### CODE REVIEW

File: contracts/erc20-gateway/ERC20Cogateway.sol [hunk 1]
```diff
@@ -14,11 +14,13 @@ pragma solidity >=0.5.0 <0.6.0;
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import "./ERC20GatewayBase.sol";
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Cogateway.sol]: This is imported twice, L17 and L19. (remove from L17)

File: contracts/erc20-gateway/ERC20Cogateway.sol [hunk 2]
```diff
@@ -30,6 +32,27 @@ contract ERC20Cogateway is
     MessageBus,
     ERC20GatewayBase {
 
+    /** Events */
+
+    /** Emitted when withdraw message is declared */
+    event WithdrawIntentDeclared(
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Cogateway.sol:37]: Can we update this as below:
```
    event WithdrawIntentDeclared(
        uint256 amount,
        uint256 nonce,
        address beneficiary,
        uint256 feeGasPrice,
        uint256 feeGasLimit,
        address withdrawer,
        address utilityToken,
        bytes32 messageHash
    );
```

I will also add a comment in the ticket.

File: contracts/erc20-gateway/ERC20Cogateway.sol [hunk 3]
```diff
@@ -88,4 +111,91 @@ contract ERC20Cogateway is
...
+     *
+     * \post Update the nonces storage mapping variable by incrementing the
+     *       value for `msg.sender` by one.
+     * \post Adds a new entry in `outbox` mapping storage variable. The value is
+     *       set as `true` for `messagehash_` in `outbox` mapping. The
+     *       `messageHash_` is calculated by `MessageOutbox.declareMessage`.
+     * \post Utility tokens will be burned from `msg.sender` account
+     *       by erc20 cogateway.
+     */
+    function withdraw(
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Cogateway.sol:152]: Should we update the params as below:

```
    function withdraw(
        uint256 _amount,
        address _beneficiary,
        uint256 _feeGasPrice,
        uint256 _feeGasLimit,
        address _utilityToken
    )
```
I will also comment it on the ticket.

File: contracts/erc20-gateway/ERC20Cogateway.sol [hunk 4]
```diff
@@ -88,4 +111,91 @@ contract ERC20Cogateway is
...
+        external
+        returns(bytes32 messageHash_)
+    {
+        require(
+            _utilityToken != address(0),
+            "Utility Token address must not be 0."
+        );
+        require(
+            _amount != 0,
+            "Withdrawal amount should be greater than 0."
```
Comment (deepesh-kn) [inline contracts/erc20-gateway/ERC20Cogateway.sol]:
```suggestion
            "Withdrawal amount should not be 0."
```

Maybe we don't need this check at all. The require at L166 is covering this check too.

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  const accountProvider = new AccountProvider(accounts);
+  let erc20Cogateway;
+  let utilityToken;
+  let setupGenesisParams = {};
+  let withdrawParam = {};
+  let withdrawIntent;
+  let sender;
+  let outboundChannelIdentifier;
+
+  const WITHDRAW_INTENT_TYPEHASH = web3.utils.keccak256(
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js:36]: `'web3' is not defined.`
How is it working?

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    "WithdrawIntent(address valueToken,address utilityToken,uint256 amount,address beneficiary)",
+  );
+
+  function hashWithdrawIntent(
+    valueToken,
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
    valueToken,
```

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+  let sender;
+  let outboundChannelIdentifier;
+
+  const WITHDRAW_INTENT_TYPEHASH = web3.utils.keccak256(
+    "WithdrawIntent(address valueToken,address utilityToken,uint256 amount,address beneficiary)",
+  );
+
+  function hashWithdrawIntent(
+    valueToken, 
+    utilityToken,
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]: `'utilityToken' is already declared in the upper scope.`

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+    amount,
+    beneficiary,
+  )
+  {
+    return web3.utils.sha3(
+      web3.eth.abi.encodeParameters(
+        ['bytes32', 'address', 'address', 'uint256', 'address'],
+        [
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
        [
```

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    utilityToken,
+    amount,
+    beneficiary,
+  )
+  {
+    return web3.utils.sha3(
+      web3.eth.abi.encodeParameters(
+        ['bytes32', 'address', 'address', 'uint256', 'address'],
+        [ 
+          WITHDRAW_INTENT_TYPEHASH,
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
          WITHDRAW_INTENT_TYPEHASH,
```

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  )
+  {
+    return web3.utils.sha3(
+      web3.eth.abi.encodeParameters(
+        ['bytes32', 'address', 'address', 'uint256', 'address'],
+        [ 
+          WITHDRAW_INTENT_TYPEHASH, 
+          valueToken,
+          utilityToken,
+          amount.toString(),
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
          amount.toString(10),
```

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    const valueToken = '0x0000000000000000000000000000000000000000';
+
+    setupGenesisParams = {
+      genesisMetachainId: Utils.generateRandomMetachainId(),
+      genesisERC20Gateway: accountProvider.get(),
+      genesisStateRootProvider: accountProvider.get(),
+      genesisMaxStorageRootItems: new BN(100),
+      genesisOutboxStorageIndex: new BN(4),
+    };
+
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion

```

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+    outboundChannelIdentifier = await erc20Cogateway.outboundChannelIdentifier.call();
+
+    withdrawIntent = await hashWithdrawIntent(
+      valueToken,
+      withdrawParam.utilityToken,
+      withdrawParam.amount,
+      withdrawParam.beneficiary,
+    );
+
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
```
`Block must not be padded by blank lines.`

File: test/erc20_gateway/erc20_cogateway/withdraw.js [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+      withdrawParam.feeGasLimit,
+      { from: sender },
+    );
+
+    const afterSenderUtilityTokenBalance = await utilityToken.balanceOf(sender);
+
+    assert.strictEqual(
+      messageHash,
+      calculatedMessageHash,
+      "Incorrect Message hash."
```
Comment (deepesh-kn) [inline test/erc20_gateway/erc20_cogateway/withdraw.js]:
```suggestion
      'Incorrect Message hash.'
```
Strings must use singlequote.

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
- Js files have linting issues.
- Added a few inline comments.
- Can you also update the documentation of `setup` function?
```
    /**
     * @notice It initializes ERC20Cogateway contract.
     *
     * \pre Setup function can be called only once.
     *
     * \post Calls `MessageOutbox.setupMessageOutbox()` with parameters
     *       `genesisMetachainId` and `genesisERC20Gateway`.
     * \post Calls `MessageInbox.setupMessageInbox` with parameters
     *       `genesisMetachainId`, `genesisERC20Gateway`,
     *       `genesisOutboxStorageIndex`, `genesisStateRootProvider` and
     *       `genesisOutboxStorageIndex`.
     */
```

File: contracts/erc20-gateway/ERC20Cogateway.sol (hunk 1 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Cogateway.sol]: Fixed.

File: contracts/erc20-gateway/ERC20Cogateway.sol (hunk 2 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Cogateway.sol:37]: Updated.

File: contracts/erc20-gateway/ERC20Cogateway.sol (hunk 3 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Cogateway.sol:152]: Updated.

File: contracts/erc20-gateway/ERC20Cogateway.sol (hunk 4 above)
Comment (Vindeep07) [reply contracts/erc20-gateway/ERC20Cogateway.sol]: Yes, require at L166 is covering same check. Done.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 10 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Done.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 12 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Fixed.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 11 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Done.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 13 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Fixed.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 7 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Done.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 6 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Fixed.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 8 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Fixed.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 9 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js]: Done.

File: test/erc20_gateway/erc20_cogateway/withdraw.js (hunk 5 above)
Comment (Vindeep07) [reply test/erc20_gateway/erc20_cogateway/withdraw.js:36]: Fixed. It was working maybe because of web3 available in truffle.

**Review by deepesh-kn — APPROVED** [review]

**Pull request merged**